analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs

```diff
@@ -33,6 +33,26 @@ public static class AnalyzerOptionsExtensions
     public static AdditionalText ProjectOutFolderPath(this AnalyzerOptions options) =>
         options.AdditionalFile("ProjectOutFolderPath.txt");
 
-    private static AdditionalText AdditionalFile(this AnalyzerOptions options, string fileName) =>
-        options.AdditionalFiles.FirstOrDefault(x => x.Path is not null && Path.GetFileName(x.Path).Equals(fileName, StringComparison.OrdinalIgnoreCase));
+    private static AdditionalText AdditionalFile(this AnalyzerOptions options, string fileName)
+    {
+        // HotPath: This code path needs to be allocation free. Don't use Linq.
+        foreach (var additionalText in options.AdditionalFiles) // Uses the struct enumerator of ImmutableArray
+        {
+            // Don't use Path.GetFilename. It allocates a string.
+            if (additionalText.Path is { } path
+                && path.EndsWith(fileName, StringComparison.OrdinalIgnoreCase))
+            {
+                // The character before the filename (if there is a character) must be a directory separator
+                if (path.Length - fileName.Length - 1 is >= 0 and var separatorPosition
```

Comment:
This line is abusing the pattern matching, making hit hard to follow. Extract the `var separatorIndex` before the `if` and then just use it the normal way.